tests/conftest.py

```diff
@@ -10,122 +11,235 @@
         message.author = author
         message.content = content
+        message.reactions = []
         return message
 
-    return _factory
+    #
+    # Helper methods to update data in useful ways.
+    #
+
+    def reset_data(self, guild_count=1, members_per_guild=1,
+                   channels_per_guild=1, messages_per_channel=1):
+        """Populates a client (the default client if not given) with test data.
+        Removes any previously configured data.
+
+        Args:
+            guild_count - number of (simulated) guilds (servers) to set up
+            members_per_guild - number of members in each simulated guild
+            channels_per_guild - number of text channels in each guild
+            messages_per_channel - number of messages in each channel's history
+        """
+
+        self.context.client.guilds[:] = []  # Erase preexisting data.
+        for gi in range(guild_count):
+            g = self.make_guild(name=f'Mock Guild {gi}')
+            self.context.client.guilds.append(g)
+
+            for mi in range(members_per_guild):
+                g.members.append(self.make_user(
+                    g, name=f'Mock Member {mi}', discriminator=1000 + mi))
+            for ci in range(channels_per_guild):
+                name = f'mock-channel-{ci}'
+                c = self.make_text_channel(g, name=name)
+                g.channels.append(c)
+                for mi in range(messages_per_channel):
+                    # Need member for message author.
+                    assert len(g.members) > 0
+                    c.test_history.append(self.make_message(
+                        g, c, g.members[mi % len(g.members)],
+                        f'Mock message {mi} in #mock-channel-{ci}'))
+
+    def queue_event(self, event_name, *args, **kwargs):
+        """Queues an event to be sent to registered listeners."""
+
+        if not event_name.startswith('on_'):
+            raise ValueError(f"event '{event_name}' doesn't start with 'on_'")
+        self.event_queue.append((event_name, args, kwargs))
+
+    async def async_dispatch_events(self):
+        """Sends all queued events to registered handlers."""
+
+        while self.event_queue:
+            batch, self.event_queue = self.event_queue, []
+            for event_name, args, kwargs in batch:
+                handler = getattr(self.context.client, event_name, None)
+                if handler is not None:
+                    await handler(*args, **kwargs)
+
+    def sim_reaction(self, message, unicode, user, delta):
+        """Simulates an emoji reaction changes and queues notification events.
+
+        Args:
+            message - the message object to modify
+            unicode - unicode of emoji to add/remove
+            user - the user adding/removing the emoji
+            delta - +1 to add, -1 to remove
+        """
+
+        if delta not in (-1, +1):
+            raise ValueError('reaction {unicode} delta {delta} not -1 or +1')
```

Comment:
Is this meant to be an f-string? (same for like 211 below)

Reply:
Good catch!